File: js/src/parse.ts
// Strict parser. Turns a NACHA text file back into the typed model without any
// positional slicing leaking into caller code. Unknown record types are ignored
// only if they are filler ('9' blocks); anything else raises.

import { RECORD_LEN, AchError } from './fields.js'

export interface ParsedAddenda {
  type: string
  info: string
  sequence: number
  entryDetailSequence: string
}
export interface ParsedEntry {
  txnCode: number
  routing: string // full 9-digit (8 + check digit)
  account: string
  amountCents: number
  idNumber: string
  name: string
  addendaFlag: boolean
  traceNumber: string
  addenda: ParsedAddenda[]
}
export interface ParsedBatch {
  serviceClass: number
  companyName: string
  companyId: string
  sec: string
  description: string
  effectiveDate: string // YYMMDD
  originDfi: string
  batchNumber: number
  entries: ParsedEntry[]
  control?: { entryAddendaCount: number; entryHash: string; totalDebit: number; totalCredit: number }
}
export interface ParsedFile {
  destinationRouting: string
  originRouting: string
  creationDate: string
  creationTime: string
  fileIdModifier: string
  batches: ParsedBatch[]
  control?: { batchCount: number; blockCount: number; entryAddendaCount: number; entryHash: string; totalDebit: number; totalCredit: number }
}

const int = (s: string) => parseInt(s.replace(/\D/g, '') || '0', 10)

export function parse(text: string): ParsedFile {
  const rows = text.split(/\r?\n/).filter((r) => r.length > 0)
  let file: ParsedFile | undefined
  let batch: ParsedBatch | undefined
  let entry: ParsedEntry | undefined

  for (const row of rows) {
    if (/^9+$/.test(row)) continue // block filler ('9' * 94), never a real record
    if (row.length !== RECORD_LEN) {
      throw new AchError('BAD_RECORD_LEN', `record is ${row.length} chars, expected ${RECORD_LEN}`)
    }
    const type = row[0]
    if (type === '1') {
      file = {
        destinationRouting: row.slice(3, 13).trim(),
        originRouting: row.slice(13, 23).trim(),
        creationDate: row.slice(23, 29),
        creationTime: row.slice(29, 33),
        fileIdModifier: row.slice(33, 34),
        batches: [],
      }
    } else if (type === '5') {
      batch = {
        serviceClass: int(row.slice(1, 4)),
        companyName: row.slice(4, 20).trim(),
        companyId: row.slice(40, 50).trim(),
        sec: row.slice(50, 53).trim(),
        description: row.slice(53, 63).trim(),
        effectiveDate: row.slice(69, 75),
        originDfi: row.slice(79, 87),
        batchNumber: int(row.slice(87, 94)),
        entries: [],
      }
      file?.batches.push(batch)
    } else if (type === '6') {
      entry = {
        txnCode: int(row.slice(1, 3)),
        routing: row.slice(3, 12),
        account: row.slice(12, 29).trim(),
        amountCents: int(row.slice(29, 39)),
        idNumber: row.slice(39, 54).trim(),
        name: row.slice(54, 76).trim(),
        addendaFlag: row[78] === '1',
        traceNumber: row.slice(79, 94),
        addenda: [],
      }
      batch?.entries.push(entry)
    } else if (type === '7') {
      entry?.addenda.push({
        type: row.slice(1, 3),
        info: row.slice(3, 83).trim(),
        sequence: int(row.slice(83, 87)),
        entryDetailSequence: row.slice(87, 94),
      })
    } else if (type === '8') {
      if (batch)
        batch.control = {
          entryAddendaCount: int(row.slice(4, 10)),
          entryHash: row.slice(10, 20),
          totalDebit: int(row.slice(20, 32)),
          totalCredit: int(row.slice(32, 44)),
        }
    } else if (type === '9') {
      if (file)
        file.control = {
          batchCount: int(row.slice(1, 7)),
          blockCount: int(row.slice(7, 13)),
          entryAddendaCount: int(row.slice(13, 21)),
          entryHash: row.slice(21, 31),
          totalDebit: int(row.slice(31, 43)),
          totalCredit: int(row.slice(43, 55)),
        }
    } else {
      throw new AchError('UNKNOWN_RECORD', `unknown record type '${type}'`)
    }
  }

  if (!file) throw new AchError('NO_FILE_HEADER', 'no file header (type 1) record found')
  return file
}

File: py/achkit/parse.py
"""Strict parser: NACHA text back into typed dicts, mirroring the JS parser."""
from __future__ import annotations

import re
from typing import Any, Dict, List

from .fields import RECORD_LEN, AchError


def _int(s: str) -> int:
    d = re.sub(r"\D", "", s)
    return int(d) if d else 0


def parse(text: str) -> Dict[str, Any]:
    rows = [r for r in text.splitlines() if r]
    file: Dict[str, Any] | None = None
    batch: Dict[str, Any] | None = None
    entry: Dict[str, Any] | None = None

    for row in rows:
        if re.fullmatch(r"9+", row):
            continue  # block filler
        if len(row) != RECORD_LEN:
            raise AchError("BAD_RECORD_LEN", f"record is {len(row)} chars, expected {RECORD_LEN}")
        t = row[0]
        if t == "1":
            file = {
                "destination_routing": row[3:13].strip(),
                "origin_routing": row[13:23].strip(),
                "creation_date": row[23:29],
                "creation_time": row[29:33],
                "file_id_modifier": row[33:34],
                "batches": [],
            }
        elif t == "5":
            batch = {
                "service_class": _int(row[1:4]),
                "company_name": row[4:20].strip(),
                "company_id": row[40:50].strip(),
                "sec": row[50:53].strip(),
                "description": row[53:63].strip(),
                "effective_date": row[69:75],
                "origin_dfi": row[79:87],
                "batch_number": _int(row[87:94]),
                "entries": [],
                "control": None,
            }
            if file:
                file["batches"].append(batch)
        elif t == "6":
            entry = {
                "txn_code": _int(row[1:3]),
                "routing": row[3:12],
                "account": row[12:29].strip(),
                "amount_cents": _int(row[29:39]),
                "id_number": row[39:54].strip(),
                "name": row[54:76].strip(),
                "addenda_flag": row[78] == "1",
                "trace_number": row[79:94],
                "addenda": [],
            }
            if batch:
                batch["entries"].append(entry)
        elif t == "7":
            if entry is not None:
                entry["addenda"].append({
                    "type": row[1:3],
                    "info": row[3:83].strip(),
                    "sequence": _int(row[83:87]),
                    "entry_detail_sequence": row[87:94],
                })
        elif t == "8":
            if batch is not None:
                batch["control"] = {
                    "entry_addenda_count": _int(row[4:10]),
                    "entry_hash": row[10:20],
                    "total_debit": _int(row[20:32]),
                    "total_credit": _int(row[32:44]),
                }
        elif t == "9":
            if file is not None:
                file["control"] = {
                    "batch_count": _int(row[1:7]),
                    "block_count": _int(row[7:13]),
                    "entry_addenda_count": _int(row[13:21]),
                    "entry_hash": row[21:31],
                    "total_debit": _int(row[31:43]),
                    "total_credit": _int(row[43:55]),
                }
        else:
            raise AchError("UNKNOWN_RECORD", f"unknown record type '{t}'")

    if file is None:
        raise AchError("NO_FILE_HEADER", "no file header (type 1) record found")
    return file
